Assets/MRTK/Core/Utilities/Physics/MixedRealityRaycaster.cs

```diff
@@ -19,96 +26,99 @@ public static bool RaycastSimplePhysicsStep(RayStep step, LayerMask[] prioritize
             return RaycastSimplePhysicsStep(step, step.Length, prioritizedLayerMasks, focusIndividualCompoundCollider, out physicsHit);
         }
 
+        private static readonly ProfilerMarker RaycastSimplePhysicsStepPerfMarker = new ProfilerMarker("[MRTK] MixedRealityRaycaster.RaycastSimplePhysicsStep");
+
         /// <summary>
         /// Simple raycasts each physics <see cref="Microsoft.MixedReality.Toolkit.Physics.RayStep"/> within a specified maximum distance.
         /// </summary>
         /// <returns>Whether or not the raycast hit something.</returns>
         public static bool RaycastSimplePhysicsStep(RayStep step, float maxDistance, LayerMask[] prioritizedLayerMasks, bool focusIndividualCompoundCollider, out RaycastHit physicsHit)
         {
-            Profiler.BeginSample("[MRTK] MixedRealityRaycaster.RaycastSimplePhysicsStep");
-
-            Debug.Assert(maxDistance > 0, "Length must be longer than zero!");
-            Debug.Assert(step.Direction != Vector3.zero, "Invalid step direction!");
-
-            bool result = prioritizedLayerMasks.Length == 1
-                // If there is only one priority, don't prioritize
-                ? UnityEngine.Physics.Raycast(step.Origin, step.Direction, out physicsHit, maxDistance, prioritizedLayerMasks[0])
-                // Raycast across all layers and prioritize
-                : TryGetPrioritizedPhysicsHit(UnityEngine.Physics.RaycastAll(step.Origin, step.Direction, maxDistance, UnityEngine.Physics.AllLayers), prioritizedLayerMasks, focusIndividualCompoundCollider, out physicsHit);
-
-            Profiler.EndSample(); // RaycastSimplePhysicsStep
-
-            return result;
+            using (RaycastSimplePhysicsStepPerfMarker.Auto())
+            {
+                Debug.Assert(maxDistance > 0, "Length must be longer than zero!");
+                Debug.Assert(step.Direction != Vector3.zero, "Invalid step direction!");
+
+                int hitCount = UnityEngine.Physics.RaycastNonAlloc(
+                    step.Origin,
+                    step.Direction,
+                    RaycastHits,
+                    maxDistance,
+                    (prioritizedLayerMasks.Length == 1) ? (int)prioritizedLayerMasks[0] : UnityEngine.Physics.AllLayers);
```

Comment:
Also realized that this leads the single layer mask case to raycast all the way to `maxDistance` instead of just the first hit object. Which also means the `TryGetPrioritizedPhysicsHit` method has to iterate through all hits to find the closest one. Are you seeing any perf differences here?

Reply:
i think i will go back to the previous single layer impl and investigate the new method in another pass